move 2d fabric (low-latency) packet header updates to sender channel

SeanNijjar opened the pull request:
This change enables an optimization for 2D (low-latency) fabric. The optimization overall removes one noc write per forwarded packet in 2D Mesh (low-latency) mode. This means for a 2D fabric-level mcast traveling in 3 directions (e.g. from west, going north, east, and south), it would save the router from having to perform three noc writes in order to process the packet.

### Old Flow

The old flow forced the packet header updates to be done by the receiver channel. The packet header update consisted of the receiver channel updating the route information in each header. This route information _had to_ be done via a noc inline write because if it was done directly in L1 before sending the full payload there would be some severe tradeoffs/options:

1. update the packet header in L1 then write and insert a noc writes sent barrier before updating the header for the next branch and forwarded packet

or

2. write to L1 and have a race if you don't have any writes sent barrier.

This extra noc write is a waste of cycles to setup the noc command. Additionally, it's yet another noc command that could be serialized/blocked behind a previous noc write; inserted stalls on `noc_cmd_buf_ready`, which is a waste of CPU time.

```
                     ┌────────────────┐                                                        
                     │                │                                                        
                     │ ┌──┐ ┌──┐ ┌──┐ │                                                        
                     │ │  │ │  │ │  │ │                                                        
                     │ │  │ │  │ │  │ │                                                        
                     │ │  │ │  │ │  │ │                                                        
                     │ └──┘ └──┘ └──┘ │                                                        
                     └───────▲────────┘                                                        
                             │                                                                 
                             │ branch 1                                                        
 ┌────────┐                  │                                                   ┌───────────┐ 
 │ inbound│                  │                   branch 2                        │           │ 
 │ ┌────┬─┼──────────────────┼───────────────────────────────────────────────────┼►┌───────┐ │ 
 │ └────┘ │                  │                                                   │ │       │ │ 
 │ rx ch  │                  │ branch 3     For Each Branch                      │ └───────┘ │ 
 └────────┘                  │                 ┌─────┬───────────────┐           │ ┌───────┐ │ 
                             │ 1. Send Packet  │hdr  │payload        │           │ │       │ │ 
                             │                 └─────┴───────────────┘           │ └───────┘ │ 
                             │                                         ┌───┐     │ ┌───────┐ │ 
                             │ 2. Update header route for this branch  │ h+│     │ │       │ │ 
                             │                                         └───┘     │ └───────┘ │ 
                             │                                                   └───────────┘ 
                             │ 3. Send credit (+1)                                             
                             │                                                                 
                     ┌───────┼────────┐                                                        
                     │       ▼        │                                                        
                     │ ┌──┐ ┌──┐ ┌──┐ │                                                        
                     │ │  │ │  │ │  │ │                                                        
                     │ │  │ │  │ │  │ │                                                        
                     │ │  │ │  │ │  │ │                                                        
                     │ └──┘ └──┘ └──┘ │                                                        
                     └────────────────┘                                                        
                                                                                               
```

### New Flow

The new flow corrects this by performing the packet header updates on the sender channel before forwarding the packet over ethernet. This is better because it removes one noc write per forwarded branch. Additionally, is also lowers the amount of work per receiver channel and spreads the load across downstream sender channels in the case of fabric multicast.

In order to accommodate this new approach, fabric unicast commands required updates to set the east/west branch offset fields in the packet header routing fields. This is so that the sender channel packet header update logic is consistent for all packet types; the sender channel doesn't need to know it's processing a unicast vs a multicast.

```                                                                  
 ┌────────┐                                                                          
 │ inbound│             ┌────────────────                     *other routers not shown*                       
 │ ┌────┬─┼─────────────┼────┐                                                       
 │ └────┘ │             │    │                                                       
 │ rx ch  │             │    │ New Flow For Each Branch (Receiver)                   
 └────────┘             │    │                                                       
                        │    │ 1. Send Packet                                        
                        │    │                                                       
                        │    │ 2. Send Credit                                        
                        │    │                                                       
                        │    │ New Flow For Each Sender Channel                      
                        │    │                                                       
                        │    │ if (is_turn):routing_fields.hop_index = sender_ch_dir 
                        │    │ else        :routing_fields.hop_index += 1            
                        │    │                                                       
                        │    │ send_packet(...)                                      
                        │    │                                                       
                     ┌──┼────┼─────────────┐                                         
                     │  ▼    ▼             │                                         
                     │ ┌──┐ ┌──┐ ┌──┐ ┌──┐ │                                         
                     │ │  │ │  │ │  │ │  │ │                                         
                     │ │  │ │  │ │  │ │  │ │                                         
                     │ │  │ │  │ │  │ │  │ │                                         
                     │ └──┘ └──┘ └──┘ └──┘ │                                         
                     │  E▲   W▲   S    N   │                                         
                     └───┼────┼────────────┘                                         
                      ┌──┴────┘                                                      
                      │                                                              
                      │                                                              
                is_turn=true                                                         
                                                                                     
```



# Performance Improvements

The performance improvements with this change are marginal overall. However, follow-up benefits of this change is that it will unblock the benefits of some other optimizations (namely the [noc hopping optimization](https://github.com/tenstorrent/tt-metal/pull/27341)) by reducing the number of back-to-back noc writes on a given cmd-buf/noc combination.

The overall improvement is as follows. Mesh shows an average improvement of around 1% with a max improvement of around 6%. 

## T3K

As is somewhat expected, this optimization is performance neutral on T3K because T3K: 
- exhibits less congestion (only one routing plane)
- Never forks 3 ways

Fabric | Speedup
-- | --
Overall | 0.9996176868
Linear | 1.000075712
Ring | 0.9999746098
Mesh LL | 0.9982428647
Mesh Dyn | 0.9999249332


## 6U

The improvement for 6u mesh is minor (at ~1% geo-mean), and a max improvement of around 6% (in that case, it's about 1% better than margin for that type of test)

Fabric | Speedup
-- | -- 
Overall | 1.002556055
Line | 0.9999918154
Ring | 1.000526053
Mesh LL | 1.011452641
Mesh Dyn | 0.9986477543

# Kernel Size Reduction

This change also reduces the overall binary size by a small amount. Dumping all the fabric router binary sizes with and without the change results in a modest binary size reduction of around 6-8%

text size new | text size old | reduction %
-- | -- | --
9496 | 10180 | 6.719056974
9508 | 10168 | 6.490952006
9208 | 9788 | 5.925623212
9336 | 9920 | 5.887096774
9208 | 9796 | 6.00244998
9084 | 9664 | 6.001655629
9120 | 9800 | 6.93877551
9120 | 9800 | 6.93877551
9760 | 10400 | 6.153846154
9496 | 10176 | 6.682389937
9508 | 10168 | 6.490952006
9028 | 9832 | 8.177379984
9120 | 9804 | 6.976744186
9748 | 10460 | 6.806883365
9764 | 10608 | 7.956259427
10048 | 10960 | 8.321167883
9584 | 10468 | 8.444784104
9596 | 10440 | 8.084291188
9512 | 10120 | 6.007905138
9840 | 10684 | 7.899663048
9496 | 10176 | 6.682389937
9080 | 9608 | 5.495420483
9208 | 9796 | 6.00244998
9120 | 9804 | 6.976744186
9644 | 10436 | 7.589114603
10136 | 10936 | 7.315288954
9080 | 9608 | 5.495420483
9084 | 9664 | 6.001655629
9756 | 10396 | 6.156213928
9792 | 10576 | 7.41301059
9496 | 10180 | 6.719056974
9208 | 9788 | 5.925623212
9756 | 10540 | 7.438330171
9892 | 10672 | 7.308845577
9336 | 9920 | 5.887096774
9512 | 10120 | 6.007905138




This change only affects the low-latency path. The routing table based routing code paths do not have this optimization applied yet; that is future work.


Issue: https://github.com/tenstorrent/tt-metal/issues/27365

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI: https://github.com/tenstorrent/tt-metal/actions/runs/17446473635 - same profiler failure on main
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI: https://github.com/tenstorrent/tt-metal/actions/runs/17326617719
- [x] Galaxy Quick: https://github.com/tenstorrent/tt-metal/actions/runs/17326765063
- [x] T3K: frequent, unit, nightly; https://github.com/tenstorrent/tt-metal/actions/runs/17326682845
  - [x] frequent, nightly 
  - [x] unit on rebased to main: https://github.com/tenstorrent/tt-metal/actions/runs/17435344507

yugaoTT commented on tt_metal/fabric/impl/kernels/edm_fabric/fabric_erisc_router.cpp:
```diff
@@ -301,6 +327,41 @@ bool did_something;
+#if defined(DYNAMIC_ROUTING_ENABLED)
+    // Unimplemented for dynamic 2D
+#else
+    static_assert(
+        my_direction == eth_chan_directions::EAST || my_direction == eth_chan_directions::WEST ||
+        my_direction == eth_chan_directions::NORTH || my_direction == eth_chan_directions::SOUTH);
```
should this assert moved to ct args file?

SeanNijjar replied:
yes I could move it there.

yugaoTT commented on tt_metal/fabric/impl/kernels/edm_fabric/fabric_erisc_router.cpp:
```diff
@@ -301,6 +327,41 @@ bool did_something;
+        is_spine_direction(eth_chan_directions::NORTH) || is_spine_direction(eth_chan_directions::SOUTH),
+        "Only spine direction of NORTH and SOUTH is supported with this code. If additional spine directions are being "
+        "added, please update the code below to support them.");
+    if constexpr (IS_FORWARDED_TRAFFIC) {
```
q: this means traffic is non-worker sent traffic?

yugaoTT commented on tt_metal/fabric/impl/kernels/edm_fabric/fabric_erisc_router.cpp:
```diff
@@ -336,7 +390,8 @@ FORCE_INLINE void send_next_data(
     }
-    internal_::eth_send_packet_bytes_unsafe(sender_txq_id, src_addr, dest_addr, payload_size_bytes);
+    internal_::eth_send_packet_bytes_unsafe(
+        sender_txq_id, reinterpret_cast<uint32_t>(pkt_header), dest_addr, payload_size_bytes);
```
what's the reason for casting here? I thought src_addr has been computed already

yugaoTT reviewed (approved):
seems like sender channel

SeanNijjar committed:
optimize fabric - remove noc write for every forwarded packet in 2d mesh

update golden t3k

fix regression for case when router is extended to mux/tensix

sq

SeanNijjar committed:
fizx bug

SeanNijjar committed:
sq

SeanNijjar committed:
update 6u
